File: LeetCode/HashMap/longest_consecutive_sequence.cpp
// https://leetcode.com/problems/longest-consecutive-sequence/?envType=study-plan-v2&envId=top-interview-150

#include <bits/stdc++.h>

using namespace std;

struct RangeData
{
    int count;
    int start;
    int end;
};

class Solution
{
public:
    int longestConsecutive(vector<int> &nums)
    {
        unordered_map<int, RangeData> mp;
        int maxCount = 0;

        for (int num : nums)
        {
            // insert the num to map if it doesn't exist
            // join on either side of range if found and update the start and end of ranges on either side

            if (mp.count(num) > 0)
            {
                continue;
            }

            int l = num - 1;
            int r = num + 1;
            RangeData rd{1, num, num};

            if (mp.count(l) > 0)
            {
                // left range exists hence merge on the left
                rd.start = mp[l].start;
                rd.count += mp[l].count;
            }

            if (mp.count(r) > 0)
            {
                // right range exists hence merge on the right
                rd.end = mp[r].end;
                rd.count += mp[r].count;
            }

            // insert the range data on the current element
            mp[num] = rd;

            if (mp.count(l) > 0)
            {
                // insert the range data on the left end
                mp[mp[l].start] = rd;
            }

            if (mp.count(r) > 0)
            {
                // insert the range data on the right end
                mp[mp[r].end] = rd;
            }

            maxCount = max(maxCount, rd.count);
        }

        return maxCount;
    }
};

int main()
{
    vector<int> v1{100, 4, 200, 1, 3, 2, 6, 0, 5, 10};

    Solution sol;

    cout << sol.longestConsecutive(v1) << endl;

    return 0;
}

File: LeetCode/HashMap/longest_consecutive_sequence.py
# https://leetcode.com/problems/longest-consecutive-sequence/?envType=study-plan-v2&envId=top-interview-150

from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        maxCount = 0
        mp = {}  # key -> [count, start_range, end_range]
        for x in nums:
            if x in mp:
                continue

            l = x - 1
            r = x + 1
            count = 1
            start = x
            end = x

            if l in mp:
                start = mp[l][1]
                count += mp[start][0]

            if r in mp:
                end = mp[r][-1]
                count += mp[end][0]

            mp[x] = [None] * 3

            mp[start][0] = count
            mp[start][-1] = end

            mp[end][0] = count
            mp[end][1] = start

            maxCount = max(maxCount, count)

        return maxCount


def main():
    v1 = [100, 4, 200, 1, 3, 2]
    v2 = [100, 4, 200, 1, 3, 2, 6, 0, 5]
    v3 = [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]

    solution = Solution()
    print(solution.longestConsecutive(v3))


main()
